# Week 3 Assignment: Pima Indians Diabetes Dataset Analysis

## I 320D: Data Science for Biomedical Informatics | Spring 2026

### 📋 Assignment Version C

---

## 🎯 This Week's Mantra

> **"Every Column Tells a Story"**

In this assignment, you'll apply the 10-Point Data Inspection to a real-world healthcare dataset focused on diabetes prediction. By the end, you'll understand not just *what* the data contains, but *why* each variable matters for clinical decision-making.

---

## Learning Objectives

By completing this assignment, you will be able to:

1. ✅ Apply the systematic 10-Point Inspection to a new healthcare dataset
2. ✅ Identify and classify feature types (continuous, discrete, categorical, ordinal)
3. ✅ Detect and document data quality issues (missing values, zero-encoded values)
4. ✅ Research and document clinical meaning for healthcare variables
5. ✅ Create meaningful data groupings based on clinical standards
6. ✅ Formulate answerable research questions about diabetes risk factors

---

## About the Dataset

**Dataset:** Pima Indians Diabetes Database  
**Source:** Originally from the National Institute of Diabetes and Digestive and Kidney Diseases  
**File:** `diabetes.csv`  
**Target Variable:** `Outcome` (whether the patient has diabetes: 1 = Yes, 0 = No)

### Clinical Context

Diabetes affects over 37 million Americans and is the 8th leading cause of death in the United States. Type 2 diabetes, which accounts for 90-95% of all diabetes cases, can often be prevented or delayed with lifestyle changes when risk factors are identified early.

This dataset contains diagnostic measurements from female patients of Pima Indian heritage, all at least 21 years old. The Pima Indians have one of the highest rates of diabetes in the world, making this population particularly important for understanding diabetes risk factors. Understanding these variables is crucial for:

- Early identification of high-risk patients
- Preventive care planning
- Resource allocation in healthcare settings
- Clinical decision support systems

---

## Getting Started

First, load the dataset and import your libraries:

In [7]:
# Import libraries
import pandas as pd
import numpy as np

# Load the dataset
df = pd.read_csv('diabetes+130-us+hospitals+for+years+1999-2008/diabetic_data.csv')

# Display first few rows to confirm it loaded
df.head()

,encounter_id,patient_nbr,race,gender,age,weight,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,...,citoglipton,insulin,glyburide-metformin,glipizide-metformin,glimepiride-pioglitazone,metformin-rosiglitazone,metformin-pioglitazone,change,diabetesMed,readmitted
0,2278392,8222157,Caucasian,Female,[0-10),?,6,25,1,1,...,No,No,No,No,No,No,No,No,No,NO
1,149190,55629189,Caucasian,Female,[10-20),?,1,1,7,3,...,No,Up,No,No,No,No,No,Ch,Yes,>30
2,64410,86047875,AfricanAmerican,Female,[20-30),?,1,1,7,2,...,No,No,No,No,No,No,No,No,Yes,NO
3,500364,82442376,Caucasian,Male,[30-40),?,1,1,7,2,...,No,Up,No,No,No,No,No,Ch,Yes,NO
4,16680,42519267,Caucasian,Male,[40-50),?,1,1,7,1,...,No,Steady,No,No,No,No,No,Ch,Yes,NO


---

## Part 1: The 10-Point Data Inspection (40 points)

Complete each inspection step and document your findings.

### Step 1: Shape (4 points)

**Your Code:**

In [8]:
print(df.shape)

(101766, 50)


**Your Findings:**
- How many rows (observations)? 101766
- How many columns (features)? 50
- What does each row represent in clinical terms? Each row represents a single hospital encounter for a diabetic patient with their demographics, diagnoses, medications, and whether they were readmitted within 30 days.

---

### Step 2: Column Names (4 points)

**Your Code:**

In [10]:
print(df.columns.tolist())

['encounter_id', 'patient_nbr', 'race', 'gender', 'age', 'weight', 'admission_type_id', 'discharge_disposition_id', 'admission_source_id', 'time_in_hospital', 'payer_code', 'medical_specialty', 'num_lab_procedures', 'num_procedures', 'num_medications', 'number_outpatient', 'number_emergency', 'number_inpatient', 'diag_1', 'diag_2', 'diag_3', 'number_diagnoses', 'max_glu_serum', 'A1Cresult', 'metformin', 'repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride', 'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide', 'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone', 'tolazamide', 'examide', 'citoglipton', 'insulin', 'glyburide-metformin', 'glipizide-metformin', 'glimepiride-pioglitazone', 'metformin-rosiglitazone', 'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted']


**Your Findings:**
- List all column names:
['encounter_id', 'patient_nbr', 'race', 'gender', 'age', 'weight', 'admission_type_id', 'discharge_disposition_id', 'admission_source_id', 'time_in_hospital', 'payer_code', 'medical_specialty', 'num_lab_procedures', 'num_procedures', 'num_medications', 'number_outpatient', 'number_emergency', 'number_inpatient', 'diag_1', 'diag_2', 'diag_3', 'number_diagnoses', 'max_glu_serum', 'A1Cresult', 'metformin', 'repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride', 'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide', 'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone', 'tolazamide', 'examide', 'citoglipton', 'insulin', 'glyburide-metformin', 'glipizide-metformin', 'glimepiride-pioglitazone', 'metformin-rosiglitazone', 'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted']

- Which columns might need further research to understand?
admission_type_id, discharge_disposition_id, admission_source_id, max_glu_serum, A1Cresult, DiabetesPedigreeFunction, and all the medication columns.
---

### Step 3: Data Types (4 points)

**Your Code:**

In [16]:
print(df.dtypes)

encounter_id                 int64
patient_nbr                  int64
race                        object
gender                      object
age                         object
weight                      object
admission_type_id            int64
discharge_disposition_id     int64
admission_source_id          int64
time_in_hospital             int64
payer_code                  object
medical_specialty           object
num_lab_procedures           int64
num_procedures               int64
num_medications              int64
number_outpatient            int64
number_emergency             int64
number_inpatient             int64
diag_1                      object
diag_2                      object
diag_3                      object
number_diagnoses             int64
max_glu_serum               object
A1Cresult                   object
metformin                   object
repaglinide                 object
nateglinide                 object
chlorpropamide              object
glimepiride         

**Your Findings:**
- Which columns are numeric?
encounter_id, patient_nbr, admission_type_id, discharge_disposition_id, admission_source_id, time_in_hospital, num_lab_procedures, num_procedures, num_medications, number_outpatient, number_emergency, number_inpatient, and number_diagnoses

- Which columns are categorical (object/string)?
race, gender, age, weight, payer_code, medical_specialty, diag_1, diag_2, diag_3, max_glu_serum, A1Cresult, metformin, repaglinide, nateglinide, chlorpropamide, glimepiride, acetohexamide, glipizide, glyburide, tolbutamide, pioglitazone, rosiglitazone, acarbose, miglitol, troglitazone, tolazamide, examide, citoglipton, insulin, glyburide-metformin, glipizide-metformin, glimepiride-pioglitazone, metformin-rosiglitazone, metformin-pioglitazone, change, diabetesMed, and readmitted

- Are there any data types that seem incorrect?
Age and weight should be integers. Also, max_glu_serum and A1Cresult include ordered clinical categories, so they should be integers.
---

### Step 4: First Look (4 points)

**Your Code:**

In [11]:
df.head()

,encounter_id,patient_nbr,race,gender,age,weight,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,...,citoglipton,insulin,glyburide-metformin,glipizide-metformin,glimepiride-pioglitazone,metformin-rosiglitazone,metformin-pioglitazone,change,diabetesMed,readmitted
0,2278392,8222157,Caucasian,Female,[0-10),?,6,25,1,1,...,No,No,No,No,No,No,No,No,No,NO
1,149190,55629189,Caucasian,Female,[10-20),?,1,1,7,3,...,No,Up,No,No,No,No,No,Ch,Yes,>30
2,64410,86047875,AfricanAmerican,Female,[20-30),?,1,1,7,2,...,No,No,No,No,No,No,No,No,Yes,NO
3,500364,82442376,Caucasian,Male,[30-40),?,1,1,7,2,...,No,Up,No,No,No,No,No,Ch,Yes,NO
4,16680,42519267,Caucasian,Male,[40-50),?,1,1,7,1,...,No,Steady,No,No,No,No,No,Ch,Yes,NO


**Your Findings:**
- What do the actual values look like?
Numeric columns like encounter_id and num_medications have clean integer values.
Categorical columns like race, gender, and age have readable string values.
Medication columns like insulin and metformin show values like No, Steady, Up, Down.
- Do you notice anything unusual or unexpected?
The age column displays ranges instead of exact numbers.
- Are there any values that look like they might be placeholders for missing data?
? is a placeholder for missing data rather than leaving cells empty (used in the weight column).
---

### Step 5: Last Look (4 points)

**Your Code:**

In [18]:
df.tail(10)

,encounter_id,patient_nbr,race,gender,age,weight,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,...,citoglipton,insulin,glyburide-metformin,glipizide-metformin,glimepiride-pioglitazone,metformin-rosiglitazone,metformin-pioglitazone,change,diabetesMed,readmitted
101756,443842070,140199494,Other,Female,[60-70),?,1,1,7,2,...,No,Steady,No,No,No,No,No,No,Yes,>30
101757,443842136,181593374,Caucasian,Female,[70-80),?,1,1,7,5,...,No,Steady,No,No,No,No,No,No,Yes,NO
101758,443842340,120975314,Caucasian,Female,[80-90),?,1,1,7,5,...,No,Up,No,No,No,No,No,Ch,Yes,NO
101759,443842778,86472243,Caucasian,Male,[80-90),?,1,1,7,1,...,No,Up,No,No,No,No,No,Ch,Yes,NO
101760,443847176,50375628,AfricanAmerican,Female,[60-70),?,1,1,7,6,...,No,Down,No,No,No,No,No,Ch,Yes,>30
101761,443847548,100162476,AfricanAmerican,Male,[70-80),?,1,3,7,3,...,No,Down,No,No,No,No,No,Ch,Yes,>30
101762,443847782,74694222,AfricanAmerican,Female,[80-90),?,1,4,5,5,...,No,Steady,No,No,No,No,No,No,Yes,NO
101763,443854148,41088789,Caucasian,Male,[70-80),?,1,1,7,1,...,No,Down,No,No,No,No,No,Ch,Yes,NO
101764,443857166,31693671,Caucasian,Female,[80-90),?,2,3,7,10,...,No,Up,No,No,No,No,No,Ch,Yes,NO
101765,443867222,175429310,Caucasian,Male,[70-80),?,1,1,7,6,...,No,No,No,No,No,No,No,No,No,NO


**Your Findings:**
- Does the data end cleanly?
Yes, the last rows don't have formatting issues.
- Are the last rows consistent with the first rows?
Yes, the same 50 columns are there with the same data: ? in the weight column, age ranges, and No, Steady, Up, and Down values.
---

### Step 6: Memory Usage (4 points)

**Your Code:**

In [12]:
print(df.memory_usage(deep=True).sum() / 1024)

225501.005859375


**Your Findings:**
- How much memory does the dataset use? 225,501 KB KB
- Is this a "small" or "large" dataset by data science standards?
Large because of the columns that contain values that are not integers.
---

### Step 7: Missing Values (4 points)

**Your Code:**

In [14]:
print(df.isnull().sum())
print()

#for the ?
for col in df.columns:
    count = (df[col] == '?').sum()
    if count > 0:
        print(f"{col}: {count} ({count/len(df)*100:.1f}%)")

encounter_id                    0
patient_nbr                     0
race                            0
gender                          0
age                             0
weight                          0
admission_type_id               0
discharge_disposition_id        0
admission_source_id             0
time_in_hospital                0
payer_code                      0
medical_specialty               0
num_lab_procedures              0
num_procedures                  0
num_medications                 0
number_outpatient               0
number_emergency                0
number_inpatient                0
diag_1                          0
diag_2                          0
diag_3                          0
number_diagnoses                0
max_glu_serum               96420
A1Cresult                   84748
metformin                       0
repaglinide                     0
nateglinide                     0
chlorpropamide                  0
glimepiride                     0
acetohexamide 

**Your Findings:**
- Which columns have missing values (according to `.isnull()`)?
max_glu_serum and A1Cresult
- What percentage of each column is missing?
race: 2.2%
weight: 96.9%
payer_code: 39.6%
medical_specialty:49.1%
diag_1: 0%
diag_2: 0.4%
diag_3: 1.4%
- ⚠️ **IMPORTANT:** This dataset uses **zeros to encode missing values** for certain physiological measurements. Which columns have zeros that are biologically impossible? (Hint: Can a living person have a blood pressure of 0? A BMI of 0? Glucose of 0?)
Since this dataset doesn't use zeros, there's a placeholder in the weight column, which overall has 97% missing values.
---

### Step 8: Duplicates (4 points)

**Your Code:**

In [15]:
print("Duplicate rows:", df.duplicated().sum())

Duplicate rows: 0


**Your Findings:**
- Are there any duplicate rows? No
- If there are duplicates, how many? N/A

---

### Step 9: Basic Statistics (4 points)

**Your Code:**

In [16]:
df.describe()

,encounter_id,patient_nbr,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,num_lab_procedures,num_procedures,num_medications,number_outpatient,number_emergency,number_inpatient,number_diagnoses
count,1.017660e+05,1.017660e+05,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000
mean,1.652016e+08,5.433040e+07,2.024006,3.715642,5.754437,4.395987,43.095641,1.339730,16.021844,0.369357,0.197836,0.635566,7.422607
std,1.026403e+08,3.869636e+07,1.445403,5.280166,4.064081,2.985108,19.674362,1.705807,8.127566,1.267265,0.930472,1.262863,1.933600
min,1.252200e+04,1.350000e+02,1.000000,1.000000,1.000000,1.000000,1.000000,0.000000,1.000000,0.000000,0.000000,0.000000,1.000000
25%,8.496119e+07,2.341322e+07,1.000000,1.000000,1.000000,2.000000,31.000000,0.000000,10.000000,0.000000,0.000000,0.000000,6.000000
50%,1.523890e+08,4.550514e+07,1.000000,1.000000,7.000000,4.000000,44.000000,1.000000,15.000000,0.000000,0.000000,0.000000,8.000000
75%,2.302709e+08,8.754595e+07,3.000000,4.000000,7.000000,6.000000,57.000000,2.000000,20.000000,0.000000,0.000000,1.000000,9.000000
max,4.438672e+08,1.895026e+08,8.000000,28.000000,25.000000,14.000000,132.000000,6.000000,81.000000,42.000000,76.000000,21.000000,16.000000


**Your Findings:**
- What is the age range in the dataset? 21 to 80
- What is the range of glucose levels? "normal" to 300
- What is the range of BMI values? underweight to obese
- Do any min values of 0 represent biologically impossible measurements?
No, the zeros in columns num_procedures and number_outpatient just mean that a patient had none of those.
---

### Step 10: Unique Counts (4 points)

**Your Code:**

In [20]:
print(df.nunique())
print(df['readmitted'].value_counts())

encounter_id                101766
patient_nbr                  71518
race                             6
gender                           3
age                             10
weight                          10
admission_type_id                8
discharge_disposition_id        26
admission_source_id             17
time_in_hospital                14
payer_code                      18
medical_specialty               73
num_lab_procedures             118
num_procedures                   7
num_medications                 75
number_outpatient               39
number_emergency                33
number_inpatient                21
diag_1                         717
diag_2                         749
diag_3                         790
number_diagnoses                16
max_glu_serum                    3
A1Cresult                        3
metformin                        4
repaglinide                      4
nateglinide                      4
chlorpropamide                   4
glimepiride         

**Your Findings:**
- Which columns have very few unique values (likely categorical or binary)?
examide, citoglipton
- Which columns have many unique values (likely continuous)?
encounter_id, patient_nbr
- What does the `Outcome` column represent and how many unique values does it have?
There are 3 unique values representing patients readmitted under 30 days, over 30 days, and not readmitted.
---

## Part 2: Data Dictionary (20 points)

Complete the following data dictionary. For each column, you must:
1. **Research** the clinical meaning
2. **Identify** the feature type (Continuous, Discrete, Categorical-Nominal, Categorical-Ordinal, Binary, Identifier)
3. **Document** the valid values/range you observe
4. **Note** any issues or questions

| Column | Description | Feature Type | Valid Values/Range | Notes/Issues |
|--------|-------------|--------------|-------------------|--------------|
| `Pregnancies` number of times pregnant | discrete, 0-17 | zero is valid
| `Glucose` | Plasma glucose concentration | Continuous | 0-199 | zeros are impossible |
| `BloodPressure` | Diastolic blood pressure (mm Hg) | continuous | 0-122 | zeros are impossible |
| `SkinThickness` | Triceps skinfold thickness (mm) | continuous | 0-99 | zeros are impossible |
| `Insulin` | 2-hour serum insulin (mu U/ml) | continuous| 0-846 | zeros are impossible |
| `BMI` | body mass index | continuous | 0-846 | zeros are impossible |
| `DiabetesPedigreeFunction` | genetic risk score based on family | continuous | 0.078 - 2.42 | higher value = greater risk |
| `Age` | patient age | discrete | 21-81 | all patients are female, 21+ |
| `Outcome` | diabetes diagnosis | binary | 0 is no, 1 is yes | target variable |

### Clinical Research Questions for Version C

Answer these questions based on your research (you may need to use Google):

**1. What is the Oral Glucose Tolerance Test (OGTT)? How is it administered and what do the glucose values in this dataset likely represent?**

Your answer: OGTT measures how well the body processes sugar. For it to be administered, the patient must fast overnight, drink a glucose solution, and have their blood sugar measured 2 hours later. The glucose is captured through max_glu_serum (the highest glucose level recorded during the hospital stay) and A1Cresult (HbA1c, which reflects average blood sugar over the past 2-3 months).

---

**2. What are the American Diabetes Association's diagnostic criteria for diabetes based on glucose levels? Include thresholds for normal, pre-diabetes, and diabetes.**

Your answer: The ADA defines normal glucose as A1C below 5.7%, pre-diabetes as A1C between 5.7–6.4%, and Ddabetes as A1C of 6.5% or higher.

---

**3. What is diastolic blood pressure (which is what's measured in this dataset)? How does hypertension relate to diabetes risk?**

Your answer: Diastolic blood pressure is the pressure in arteries between heartbeats. Hypertension and diabetes commonly co-occur as part of metabolic syndrome.

---

**4. What is the "metabolic syndrome" and how do the variables in this dataset relate to its diagnostic criteria?**

Your answer: Metabolic syndrome is a cluster of high blood sugar, excess body fat, high blood pressure, and abnormal cholesterol, which together increase the risk of heart disease and diabetes. The variables max_glu_serum (blood sugar), weight (body fat, though mostly missing), insulin (glucose regulation), and diabetesMed (treatment status) relate to metabolic syndrome.

---

## Part 3: Data Validation (15 points)

### 3.1 Zero-Value Analysis (5 points)

A unique challenge in this dataset is that **zeros are used to encode missing values** for several columns where zero is biologically impossible.

Write code to count how many zeros exist in each column:

In [21]:
for col in df.columns:
    count = (df[col] == '?').sum()
    if count > 0:
        pct = count/len(df)*100
        print(f"{col}: {count} ({pct:.1f}%)")

race: 2273 (2.2%)
weight: 98569 (96.9%)
payer_code: 40256 (39.6%)
medical_specialty: 49949 (49.1%)
diag_1: 21 (0.0%)
diag_2: 358 (0.4%)
diag_3: 1423 (1.4%)


**Your Findings:**

- Which columns have zeros that are biologically impossible?
Glucose, BloodPressure, BMI, and SkinThickness have zeros which is impossible
- For each problematic column, what percentage of values are zeros?
weight: 96.9%
medical_specialty: 49.1%
payer_code: 39.6%
race: 2.2%
diag_1: 0.0%
diag_2: 0.4%
diag_3: 1.4%
- Why would the original data collectors use zeros instead of leaving these as empty/null values?
They used ? as placeholders in place of true null values
---

### 3.2 Glucose Validation (5 points)

Write code to examine glucose values. Check for zeros and understand the distribution.

**Your Code:**

In [22]:
print(df['max_glu_serum'].value_counts(dropna=False))
print()

# Count missing
missing = df['max_glu_serum'].isna().sum()
total = len(df)
print(f"Missing glucose values: {missing} ({missing/total*100:.1f}%)")
print()

# Distribution excluding missing
print(df['max_glu_serum'].value_counts())

max_glu_serum
NaN     96420
Norm     2597
>200     1485
>300     1264
Name: count, dtype: int64

Missing glucose values: 96420 (94.7%)

max_glu_serum
Norm    2597
>200    1485
>300    1264
Name: count, dtype: int64


**Your Findings:**

- How many patients have a glucose value of 0?
96,420
- What is the valid glucose range in this dataset (excluding zeros)?
normal (2597), >200 (1485), >300 (1264)
- Based on the maximum glucose value, do all values seem clinically plausible for a 2-hour OGTT?
Yes
---

### 3.3 SkinThickness Validation (5 points)

Write code to examine skin thickness values and identify problematic entries.

**Your Code:**

In [28]:
df = pd.read_csv('diabetes+130-us+hospitals+for+years+1999-2008/diabetes.csv')
st = df['SkinThickness']
print(st.describe())

# Q1: Count zero values
zero_count = (st == 0).sum()
print(f"\nPatients with SkinThickness = 0: {zero_count}")



count    768.000000
mean      20.536458
std       15.952218
min        0.000000
25%        0.000000
50%       23.000000
75%       32.000000
max       99.000000
Name: SkinThickness, dtype: float64

Patients with SkinThickness = 0: 227
This represents 29.6% of the dataset


**Your Findings:**

- How many patients have a skin thickness value of 0?
227
- What is a typical range for triceps skinfold thickness in adults? (Research this)
16-25mm for females
- Are there any non-zero values that seem clinically unlikely?
Row 445 that said 63mm and 579 that said 99mm
---

## Part 4: Create Clinical Glucose Groups (10 points)

Create a new column called `glucose_category` that categorizes patients into clinically-meaningful groups.

### Version C: Glycemic Status Categories

Use these categories based on the American Diabetes Association diagnostic criteria for the 2-hour Oral Glucose Tolerance Test:

| Glucose Category | Range (mg/dL) | Clinical Rationale |
|------------------|---------------|-------------------|
| Normal | < 140 | Normal glucose tolerance |
| Pre-diabetes | 140-199 | Impaired glucose tolerance, high risk for developing diabetes |
| Diabetic Range | 200+ | Meets diagnostic criteria for diabetes |

**⚠️ Important:** You must handle Glucose values of 0 (missing data) before creating categories. Either exclude them or create a separate "Missing" category.

### Your Code:

In [31]:
# First, handle the zero Glucose values
df['Glucose'] = df['Glucose'].replace(0, pd.NA)
def categorize_glucose(glucose):
    if pd.isna(glucose):
        return 'Missing'
    elif glucose < 140:
        return 'Normal'
    elif glucose <= 199:
        return 'Pre-diabetes'
    else:
        return 'Diabetic Range'


Glucose Category Counts:
glucose_category
Normal          566
Pre-diabetes    197
Missing           5
Name: count, dtype: int64

Sample rows:
  Glucose glucose_category
0     148     Pre-diabetes
1      85           Normal
2     183     Pre-diabetes
3      89           Normal
4     137           Normal
5     116           Normal
6      78           Normal
7     115           Normal
8     197     Pre-diabetes
9     125           Normal


### Verify your groupings worked:

In [32]:
# Show counts per glucose category
print("\nGlucose Category Counts:")
print(df['glucose_category'].value_counts())

print("\nSample rows:")
print(df[['Glucose', 'glucose_category']].head(10))

df['glucose_category'] = df['Glucose'].apply(categorize_glucose)


Glucose Category Counts:
glucose_category
Normal          566
Pre-diabetes    197
Missing           5
Name: count, dtype: int64

Sample rows:
  Glucose glucose_category
0     148     Pre-diabetes
1      85           Normal
2     183     Pre-diabetes
3      89           Normal
4     137           Normal
5     116           Normal
6      78           Normal
7     115           Normal
8     197     Pre-diabetes
9     125           Normal


### Calculate diabetes rate by glucose category:

In [36]:
# Handle zeros and create glucose_category (from previous section)
df['Glucose'] = df['Glucose'].replace(0, pd.NA)

def categorize_glucose(glucose):
    if pd.isna(glucose):
        return 'Missing'
    elif glucose < 140:
        return 'Normal'
    elif glucose <= 199:
        return 'Pre-diabetes'
    else:
        return 'Diabetic Range'

df['glucose_category'] = df['Glucose'].apply(categorize_glucose)

# ── Exclude missing glucose patients ──────────────────────────────
df_clean = df[df['glucose_category'] != 'Missing']

# ── Calculate diabetes percentage per category ────────────────────
diabetes_rate = (
    df_clean.groupby('glucose_category')['Outcome']
    .agg(
        Total='count',
        Diabetic='sum'
    )
    .assign(Diabetes_Pct=lambda x: (x['Diabetic'] / x['Total'] * 100).round(1))
)

print("Diabetes Rate by Glucose Category:")
print(diabetes_rate)

Diabetes Rate by Glucose Category:
                  Total  Diabetic  Diabetes_Pct
glucose_category                               
Normal              566       131          23.1
Pre-diabetes        197       135          68.5


### Analysis Questions:

**1. How many patients are in each glucose category (excluding missing)?**

Your answer: Normal: 566 patients, Pre-diabetes: 197 patients, Diabetic range: 0 patients

---

**2. What is the diabetes rate (percentage with Outcome=1) for each glucose category?**

Your answer: Normal: 23.1%, pre-diabetes: 68.5%

---

**3. You might notice something surprising: some patients with glucose in the "Diabetic Range" (≥200) have Outcome=0 (no diabetes), and some with "Normal" glucose have Outcome=1 (diabetes). What might explain these apparent contradictions?**

Your answer: Other factors affect diabetes in addition to glucose levels, treatment effects, and how it's only one glucose measurement.

---

## Part 5: Research Questions (15 points)

### 5.1 Write Three Answerable Questions (9 points)

Write three questions that THIS dataset can answer. Remember: the data can show relationships and patterns, but cannot prove causation.

**Your questions must explore these specific areas:**

1. **A question about glucose levels and diabetes outcome:**

Is there a statistically significant difference in glucose levels between patients who have diabetes and those who do not, and how strong is that association?

2. **A question comparing patients by blood pressure levels:**

Do patients with a BMI classified as obese have a higher rate of diabetes diagnosis compared to patients with a normal BMI, and is this relevant is across different age groups?

3. **A question about the combination of insulin AND BMI:**

Among patients with pre-diabetes glucose levels, which additional risk factors are most strongly correlated with a diabetes diagnosis?

### 5.2 Identify One Question the Data CANNOT Answer (3 points)

Write one question about **HbA1c (glycated hemoglobin) or long-term glucose control** that this dataset cannot answer, and explain why.

**Question

Does long-term poor glycemic control (measured by HbA1c) predict diabetes in Pima Indian women better than a single fasting glucose reading?

**Why it cannot be answered with this data:**
This dataset contains no HbA1c column. HbA1c reflects average blood glucose over the past 2–3 months. The only glucose-related measurement available is a single plasma glucose reading, which is a snapshot in time and can fluctuate day to day. To answer this question you would need a dataset that includes both an HbA1c reading and a fasting glucose reading for the same patients, then compare how well each one predicts the Outcome variable.


### 5.3 Grouping Analysis (3 points)

Answer this question using a groupby analysis:

**"What is the average BMI for patients with diabetes (Outcome=1) versus without diabetes (Outcome=0)?"**

(Note: You should exclude zero BMI values before calculating!)

**Your Code:**

In [37]:
print(f"Zero BMI values found: {(df['BMI'] == 0).sum()}")
df_clean = df[df['BMI'] != 0]

# ── Groupby analysis: average BMI by diabetes outcome ─────────────
bmi_by_outcome = (
    df_clean.groupby('Outcome')['BMI']
    .agg(
        Count='count',
        Mean_BMI='mean',
        Median_BMI='median',
        Std_BMI='std'
    )
    .round(2)
)

# Rename index for readability
bmi_by_outcome.index = ['No Diabetes (0)', 'Diabetes (1)']

print("\nAverage BMI by Diabetes Outcome:")
print(bmi_by_outcome)

Zero BMI values found: 11

Average BMI by Diabetes Outcome:
                 Count  Mean_BMI  Median_BMI  Std_BMI
No Diabetes (0)    491     30.86        30.1     6.56
Diabetes (1)       266     35.41        34.3     6.61


**Your Interpretation:**

What is the difference in average BMI between diabetic and non-diabetic patients? Is this difference clinically meaningful?

35.14 − 30.30 = 4.84 BMI units

Yes, it's meaningful. The two groups fall into different clinical categories altogether. BMI is an imperfect measure so it should be interpreted alongside the other factors.

---

## Part 6: Target Variable Analysis (Bonus - 5 points)

The `Outcome` column is our **target variable** - what we're trying to predict. Analyze its distribution.

**Your Code:**

In [38]:
# Show the count and percentage of diabetic vs non-diabetic patients
outcome_counts = df['Outcome'].value_counts()
outcome_pct    = df['Outcome'].value_counts(normalize=True) * 100

summary = pd.DataFrame({
    'Label'      : ['No Diabetes', 'Diabetes'],
    'Count'      : outcome_counts.values,
    'Percentage' : outcome_pct.round(1).values
})

print(summary.to_string(index=False))

      Label  Count  Percentage
No Diabetes    500        65.1
   Diabetes    268        34.9


### Bonus Questions:

**1. What percentage of patients in this dataset have diabetes?**

Your answer: 34.9%

---

**2. Is this dataset balanced or imbalanced? (A balanced dataset has roughly equal classes)**

Your answer: Imbalanced because it's about a 2:1 ratio in favor of non-diabetic patients.

---

**3. Why does class imbalance matter for machine learning? (You may need to research this)**

Your answer: Most ML algorithms assume classes are  equal and optimize for overall accuracy. With a 65/35 split, a model could predict "No Diabetes" for every single patient and still achieve 65% accuracy. The accuracy paradox is that imbalanced data tends to make models biased toward the majority class, causing them to miss the minority class more often. In medicine, a false negative is far more harmful than a false positive.

---

**4. The Pima Indians have one of the highest rates of diabetes in the world (around 50% in some studies). How does this dataset's diabetes rate compare? What might this tell you about the patient selection criteria?**

Your answer: This dataset shows that about 77% of patients are on diabetes medication, and since all patients in this dataset were admitted to hospital with diabetes-related encounters, the "rate" is 100%. Patient selection criteria are extremely important, especially based on location, but this cannot be generalized.

---

## Submission Checklist

Before submitting, verify you have completed:

- [ ] **Part 1:** All 10 inspection steps with code AND written findings
- [ ] **Part 2:** Complete data dictionary with all 9 columns filled in
- [ ] **Part 2:** Answered all 4 clinical research questions
- [ ] **Part 3:** All 3 validation checks with code and answers
- [ ] **Part 4:** Created `glucose_category` column using the **Glycemic Status Categories**
- [ ] **Part 4:** Calculated diabetes rate by glucose category with interpretation
- [ ] **Part 5:** Three research questions (glucose, blood pressure, insulin+BMI)
- [ ] **Part 5:** One unanswerable question about HbA1c/long-term glucose control
- [ ] **Part 5:** BMI by outcome groupby analysis
- [ ] **Bonus (Optional):** Target variable analysis

---

## Grading Rubric

| Component | Points | Requirements for Full Credit |
|-----------|--------|------------------------------|
| Part 1: 10-Point Inspection | 40 | All 10 steps complete with working code AND thoughtful written analysis |
| Part 2: Data Dictionary | 20 | All 9 columns documented with correct feature types and clinical research |
| Part 3: Data Validation | 15 | All validation checks complete with working code and insightful answers |
| Part 4: Glucose Categories | 10 | Working code that creates correct groups, handles missing data, AND meaningful interpretation |
| Part 5: Research Questions | 15 | Three good questions in specified areas, one clear limitation, groupby analysis complete |
| **Bonus:** Target Analysis | +5 | Thoughtful analysis of class imbalance with real-world connection |
| **Total** | 100 (+5 bonus) | |

---

## Hints (Read Before You Get Stuck!)

### ⚠️ Common Pitfalls:

1. **Zeros as Missing Values**
   - Unlike most datasets, this one uses **0** to represent missing values for columns like `Glucose`, `BloodPressure`, `SkinThickness`, `Insulin`, and `BMI`
   - The `.isnull()` method won't detect these!
   - You need to check for zeros manually in columns where 0 is biologically impossible

2. **Glucose of 0 will break your categories** - you MUST handle these before using `pd.cut()` or your analysis will be incorrect

3. **The glucose measurement is from a 2-hour OGTT** - not a fasting glucose test. The thresholds are different!

4. **Outcome vs. Glucose Category mismatch** - Don't be surprised if they don't perfectly align. Think about why this might be.

### 💡 Pro Tips:

- Use `value_counts()` liberally to understand distributions
- Use `df[df['ColumnName'] == 0].shape[0]` to count zeros in a column
- When creating glucose categories, consider using `df[df['Glucose'] > 0]` to filter first
- Use `pd.cut()` with appropriate bin edges: `[0, 140, 200, float('inf')]`
- The `describe()` function is your friend - look at those min values!

---

## Useful Resources

- **Original Dataset Source:** UCI Machine Learning Repository
- **Diabetes Diagnostic Criteria:** https://www.diabetes.org/a1c/diagnosis
- **Oral Glucose Tolerance Test:** https://www.mayoclinic.org/tests-procedures/glucose-tolerance-test/about/pac-20394296
- **Metabolic Syndrome:** https://www.heart.org/en/health-topics/metabolic-syndrome/about-metabolic-syndrome
- **Pandas Documentation:** https://pandas.pydata.org/docs/

---

*Remember: "Every Column Tells a Story" - your job is to figure out what that story is!*

---

**Due Date:** [See Canvas]

**Submission:** Upload your completed Jupyter notebook (.ipynb) to Canvas